### PR TITLE
[Structural] Add GetSpecifications to some elements

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**📝 Description**
In this PR I'm adding the `GetSpecifications` method to some of the elements in the `StructuralMechanicsApplication`. I did it in such a way that we can retrieve the DOFs from the `required_dofs` field. That is to say, not to all the elements but to "all" the base elements. Besides this, I may have missed some of the fields so I'd appreciate any comment from you @loumalouomega @philbucher guys.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 1]
```diff
@@ -971,6 +971,36 @@ int CrBeamElement2D2N::Check(const ProcessInfo& rCurrentProcessInfo) const
...
+        },
+        "required_variables"         : ["DISPLACEMENT","ROTATION"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","ROTATION_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line2D2"],
+        "element_integrates_in_time" : false,
+        "compatible_constitutive_laws": {
+            "type"        : ["BeamConstitutiveLaw"],
+            "dimension"   : ["2D"],
+            "strain_size" : [1]
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]: @clazaro and @KlausBSautter for a 2D beam the strain size should be 2 (sigma_xx, tau_xy) and in 3D should be 3 (sigma_xx, tau_xy, tau_xz) ?

In any case we should also discuss in general the meaning of the dimension for a CL (are shells 3D¿?, are trusses 1D¿?)

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp [hunk 2]
```diff
@@ -1726,6 +1726,36 @@ void CrBeamElement3D2N::FinalizeNonLinearIteration(const ProcessInfo& rCurrentPr
...
+            "entity"                 : []
+        },
+        "required_variables"         : ["DISPLACEMENT","ROTATION"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z","ROTATION_X","ROTATION_Y","ROTATION_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line3D2"],
+        "element_integrates_in_time" : false,
+        "compatible_constitutive_laws": {
+            "type"        : ["BeamConstitutiveLaw"],
+            "dimension"   : ["3D"],
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:1749]: I am confused with the dimensions and strain sizes, @KratosMultiphysics/structural-mechanics we should discuss this

File: applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp [hunk 3]
```diff
@@ -1315,6 +1315,36 @@ void MembraneElement::CalculateDampingMatrix(
...
+            "entity"                 : []
+        },
+        "required_variables"         : ["DISPLACEMENT","ROTATION"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z","ROTATION_X","ROTATION_Y","ROTATION_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Triangle3D3", "Quadrilateral3D4"],
+        "element_integrates_in_time" : false,
+        "compatible_constitutive_laws": {
+            "type"        : ["PlaneStress"],
+            "dimension"   : ["2D"],
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp]: Again, 3D¿?

**Review by loumalouomega — CHANGES_REQUESTED** [review]
We should discuss the dimensions and strains sizes of the slender elements first @KratosMultiphysics/structural-mechanics

File: applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp (hunk 3 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp]: Changed.

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 4]
```diff
@@ -971,6 +971,36 @@ int CrBeamElement2D2N::Check(const ProcessInfo& rCurrentProcessInfo) const
...
+        },
+        "required_variables"         : ["DISPLACEMENT","ROTATION"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","ROTATION_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line2D2"],
+        "element_integrates_in_time" : false,
+        "compatible_constitutive_laws": {
+            "type"        : ["BeamConstitutiveLaw"],
+            "dimension"   : ["1D"],
+            "strain_size" : [1]
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]:
```suggestion
            "strain_size" : [2]

```

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp [hunk 5]
```diff
@@ -1726,6 +1726,36 @@ void CrBeamElement3D2N::FinalizeNonLinearIteration(const ProcessInfo& rCurrentPr
...
+        },
+        "required_variables"         : ["DISPLACEMENT","ROTATION"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z","ROTATION_X","ROTATION_Y","ROTATION_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line3D2"],
+        "element_integrates_in_time" : false,
+        "compatible_constitutive_laws": {
+            "type"        : ["BeamConstitutiveLaw"],
+            "dimension"   : ["1D"],
+            "strain_size" : [1]
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]:
```suggestion
            "strain_size" : [3]

```

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp [hunk 6]
```diff
@@ -1726,6 +1726,37 @@ void CrBeamElement3D2N::FinalizeNonLinearIteration(const ProcessInfo& rCurrentPr
...
+            "entity"                 : []
+        },
+        "required_variables"         : ["DISPLACEMENT","ROTATION"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z","ROTATION_X","ROTATION_Y","ROTATION_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line3D2"],
+        "element_integrates_in_time" : false,
+        "compatible_constitutive_laws": {
+            "type"        : ["BeamConstitutiveLaw"],
+            "dimension"   : ["1D"],
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: Still confused about this... (we can discuss later)

**Review by loumalouomega — DISMISSED** [review]
Okay

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp (hunk 1 above)
Comment (clazaro) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]: Disclaimer: still a Kratos newbie... That said, the Beam Constitutive Law is an empty law: the constitutive behavior is defined within the element code and terms of generalized stresses. The results at integration points are: (N, V2, M3) in the case of the 2D beam, and (N, V2, V3, Mt, M2, M3) in the case of the 3D beam. There is no evaluation of stresses at material points of the cross section.

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp (hunk 6 above)
Comment (clazaro) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: IMO, the meaning is clear when dealing with (non-generalized) strains/stresses:
Uniaxial --> dimension = 1D --> size = 1 (sigma_xx)
Planar --> dimension = 2D --> size = 3 (sigma_xx, sigma_yy, tau_xy)
and the 3D case with size = 6.

The case of the 2D and 3D beams is different, as explained in my previous comment: they use generalized strains/stresses for the constitutive definition.

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]: So shall we keep 1 as it is right now?

Comment (clazaro) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]: I'd vote for:

2D beam: CL dimension = 2 ; strain_size = 3
3D beam: CL dimension = 3 ; strain_size = 6

Although the meaning of 'dimension' is just inherited from the embedding space.

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]: > Although the meaning of 'dimension' is just inherited from the embedding space.

Thanks @clazaro, I was thinking this

> 2D beam: CL dimension = 2 ; strain_size = 3
> 3D beam: CL dimension = 3 ; strain_size = 6

For the strain size I was thinking in the neutral axis, but material points make more sense

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 7]
```diff
@@ -971,6 +971,37 @@ int CrBeamElement2D2N::Check(const ProcessInfo& rCurrentProcessInfo) const
...
+            "entity"                 : []
+        },
+        "required_variables"         : ["DISPLACEMENT","ROTATION"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","ROTATION_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line2D2"],
+        "element_integrates_in_time" : false,
+        "compatible_constitutive_laws": {
+            "type"        : ["BeamConstitutiveLaw"],
+            "dimension"   : ["1D"],
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]:
```suggestion
            "dimension"   : ["2D"],

```

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 8]
```diff
@@ -971,6 +971,37 @@ int CrBeamElement2D2N::Check(const ProcessInfo& rCurrentProcessInfo) const
...
+            "strain_size" : [2]
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]:
```suggestion
            "strain_size" : [3]

```

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp (hunk 6 above)
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]:
```suggestion
            "dimension"   : ["3D"],

```

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp [hunk 9]
```diff
@@ -1726,6 +1726,37 @@ void CrBeamElement3D2N::FinalizeNonLinearIteration(const ProcessInfo& rCurrentPr
...
+            "strain_size" : [3]
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]:
```suggestion
            "strain_size" : [6]

```

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp [hunk 10]
```diff
@@ -983,6 +983,36 @@ bool TrussElement3D2N::HasSelfWeight() const
...
+            "entity"                 : []
+        },
+        "required_variables"         : ["DISPLACEMENT"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line3D2"],
+        "element_integrates_in_time" : false,
+        "compatible_constitutive_laws": {
+            "type"        : ["TrussConstitutiveLaw"],
+            "dimension"   : ["1D"],
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]:
```suggestion
            "dimension"   : ["3D"],
```

**Review by loumalouomega — DISMISSED** [review]
Okay now

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]: Thanks  @clazaro. I'll change it accordingly.

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**